bpf/kmesh/workload/sockops_tuple.c

```diff
@@ -14,104 +14,178 @@
+	return 0;
 }
 
-// insert an IPv4 tuple into the ringbuf, return 0 if succeed, 1 if failed
-static inline int insert_ipv4_tuple(struct bpf_sock_ops *skops)
+static inline void enable_encoding_metadata(struct bpf_sock_ops *skops, struct bpf_sock_tuple *tuple_info)
 {
-    if (skops->family == AF_INET) {
-        struct ringbuf_msg_type *msg = bpf_ringbuf_reserve(&map_of_tuple, sizeof(*msg), 0);
-        if (!msg) {
-            BPF_LOG(WARN, SOCKOPS, "can not alloc new ringbuf in map_of_tuple");
-          return 1;
-        }
-        (*msg).tuple.ipv4.daddr = skops->local_ip4;
-        (*msg).tuple.ipv4.saddr = skops->remote_ip4;
-        // local_port is host byteorder
-        (*msg).tuple.ipv4.dport = bpf_htonl(skops->local_port) >> 16;
-        // remote_port is network byteorder
-        (*msg).tuple.ipv4.sport = skops->remote_port >> 16;
-        (*msg).type = (__u32)IPV4;
-        bpf_ringbuf_submit(msg, 0);
-    }
-    return 0;
+	int err;
+	err = bpf_sock_hash_update(skops, &map_of_kmesh_hashmap, tuple_info, BPF_ANY);
+	if (err)
+		BPF_LOG(ERR, SOCKOPS, "enable encoding metadta failed!, err is %d\n", err);
 }
 
 SEC("sockops")
 int record_tuple(struct bpf_sock_ops *skops)
 {
-    if(!dst_is_managed(skops)) {
-        return 0;
-    }
-
-    switch (skops->op) {
-        case BPF_SOCK_OPS_ACTIVE_ESTABLISHED_CB:
-            if(bpf_sock_ops_cb_flags_set(skops, BPF_SOCK_OPS_STATE_CB_FLAG) != 0){
-                BPF_LOG(ERR, SOCKOPS, "set sockops cb failed!\n");
-            }
-            break;
-        case BPF_SOCK_OPS_PASSIVE_ESTABLISHED_CB:
-            if(bpf_sock_ops_cb_flags_set(skops, BPF_SOCK_OPS_STATE_CB_FLAG) != 0){
-                BPF_LOG(ERR, SOCKOPS, "set sockops cb failed!\n");
-            }           
-            if(insert_ipv4_tuple(skops)) {
-                break;
-            }
-        case BPF_SOCK_OPS_STATE_CB:
-            if(skops->args[1] == BPF_TCP_CLOSE || skops->args[1] == BPF_TCP_CLOSE_WAIT 
-            || skops->args[1] == BPF_TCP_FIN_WAIT1) {
-                clean_auth_map(skops);
-            }
-            // not support IPV6
-            break;
-        default:
-            break;
-    }
-    return 0;
+	// not support IPV6
+	if (skops->family != AF_INET)
+		return 0;
+	struct bpf_sock_tuple tuple_info = {0};
+	extract_key4_from_ops(skops, &tuple_info);
+	// in ACCTIVE ESTABLISHED CB state
+	// sender check src(client ip) is managed by the kmesh. 
+	// If yes, metadata encoding is required.
+	// in PASSIVE ESTABLISHED CB state
+	// reciver check src(server ip) is managed by the kmesh.
+	// If yes, auth need to be check.
+	// NOTICE! in PASSIVE ESTABLISHED CB state, server ip is src ip!
+	if (!kmesh_manager_by_ip(tuple_info.ipv4.saddr))
+		return 0;
+	switch (skops->op) {
+		case BPF_SOCK_OPS_ACTIVE_ESTABLISHED_CB:
+			if(bpf_sock_ops_cb_flags_set(skops, BPF_SOCK_OPS_STATE_CB_FLAG) != 0)
+				BPF_LOG(ERR, SOCKOPS, "set sockops cb failed!\n");
+			enable_encoding_metadata(skops, &tuple_info);
```

Comment:
This seems not right, only when we have waypoint, we need to encode. Here we cannot know if we have any waypoint, right?